core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -202,6 +202,13 @@ class Partition(val topic: String,
           true
         }
       val leaderReplica = getReplica().get
+      val curLeaderLogEndOffset = leaderReplica.logEndOffset.messageOffset
+      val curTimeMs = time.milliseconds
+      // initialize lastCaughtUpTime of replicas as well as their lastFetchTimeMs and lastFetchLeaderLogEndOffset.
+      (assignedReplicas() - leaderReplica).foreach{replica =>
+        val lastCaughtUpTimeMs = if (inSyncReplicas.contains(replica)) curTimeMs else 0L
+        replica.resetLastCaughtUpTime(curLeaderLogEndOffset, curTimeMs, lastCaughtUpTimeMs)
+      }
```

Comment:
1. I think we need to call `resetLastCaughtUpTime()` even if isNewLeader is false. For example, say a follower is fully caught up before it is offline. If the leader doesn't reset lastCaughtUpTime of this follower, then in the next ReplicaLagTimeMaxMs the replica will be in ISR and hw of this partition can not increase, which essentially makes this replica unavailable to producers.

2. I think we should initialize lastFetchLeaderLogEndOffset to current LEO of the leader and lastFetchTimeMs to current time regardless of whether the follower is in ISR. Although the variables are named as `lastFetch*` and updated when the fetch request is received, they are actually used to record the latest snapshot of ( leader LEO, time) pair. They should be interpreted in this way: the leader's LEO is `x` at time `t`. If follower's LEO >= x, then its lastCaughtTimeMs >= t. Given this interpretation, it probably doesn't make sense to set `lastFetchLeaderLogEndOffset = MAX_LONG and lastFetchTimeMs = 0`.